src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp

```diff
@@ -72,6 +61,53 @@ bool SmokeCoAlarmServer::SetExpressedState(EndpointId endpointId, ExpressedState
     return success;
 }
 
+void SmokeCoAlarmServer::SetExpressedStateByPriority(EndpointId endpointId,
+                                                     const std::array<ExpressedStateEnum, kPriorityOrderLength> & priorityOrder)
+{
+    AlarmStateEnum alarmState          = AlarmStateEnum::kNormal;
+    EndOfServiceEnum endOfServiceState = EndOfServiceEnum::kNormal;
+    bool active                        = false;
+
+    for (ExpressedStateEnum priority : priorityOrder)
+    {
+        switch (priority)
+        {
+        case ExpressedStateEnum::kSmokeAlarm:
+            VerifyOrReturn(GetSmokeState(endpointId, alarmState));
+            break;
+        case ExpressedStateEnum::kCOAlarm:
+            VerifyOrReturn(GetCOState(endpointId, alarmState));
+            break;
+        case ExpressedStateEnum::kBatteryAlert:
+            VerifyOrReturn(GetBatteryAlert(endpointId, alarmState));
+            break;
+        case ExpressedStateEnum::kTesting:
+            VerifyOrReturn(GetTestInProgress(endpointId, active));
+            break;
+        case ExpressedStateEnum::kHardwareFault:
+            VerifyOrReturn(GetHardwareFaultAlert(endpointId, active));
+            break;
+        case ExpressedStateEnum::kEndOfService:
+            VerifyOrReturn(GetEndOfServiceAlert(endpointId, endOfServiceState));
+            break;
+        case ExpressedStateEnum::kInterconnectSmoke:
+            VerifyOrReturn(GetInterconnectSmokeAlarm(endpointId, alarmState));
+            break;
+        case ExpressedStateEnum::kInterconnectCO:
+            VerifyOrReturn(GetInterconnectCOAlarm(endpointId, alarmState));
+            break;
+        default:
+            break;
+        }
+
+        VerifyOrReturn(alarmState == AlarmStateEnum::kNormal, SetExpressedState(endpointId, priority));
```

Comment:
you are losing the CHIP_ERROR in the setter though.

I would use some LogErrorOnFailure calls:

```cpp
  if ((alarmState != AlarmStateEnum::kNormal) || (endOfServiceState != EndOfServiceEnum::kNormal) || active) {
     LogErrorOnFailure(SetExpressedState(endpointId, priority));
     return;
  }

// ....
LogErrorOnFailure(SetExpressedState(endpointId, ExpressedStateEnum::kNormal));
```

Reply:
VerifyOrDo and ChipLogError are used for now, will be improved in a later PR.